src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -309,47 +309,90 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
+            Solution solution,
+            ProjectId projectId,
+            Dictionary<ProjectId, ArrayBuilder<ProjectState>> projectIdToSiblingFlavors,
+            [NotNullWhen(true)] out ArrayBuilder<ProjectState>? siblingFlavors)
+        {
+            siblingFlavors = null;
+
+            var projectState = solution.SolutionState.GetRequiredProjectState(projectId);
+
+            // If this project doesn't have any flavors itself, then there's no sibling flavors of it to return.
+            if (projectState.NameAndFlavor.flavor is null)
+                return false;
+
+            if (!projectIdToSiblingFlavors.TryGetValue(projectId, out siblingFlavors))
+            {
+                siblingFlavors = ArrayBuilder<ProjectState>.GetInstance();
+                projectIdToSiblingFlavors.Add(projectId, siblingFlavors);
+
+                foreach (var (siblingProjectId, siblingProject) in solution.SolutionState.ProjectStates)
+                {
+                    if (projectId == siblingProjectId)
+                        continue;
+
+                    if (siblingProject.NameAndFlavor.name == projectState.NameAndFlavor.name &&
+                        siblingProject.NameAndFlavor.flavor != null)
+                    {
+                        siblingFlavors.Add(siblingProject);
+                    }
+                }
+            }
+
+            return true;
+        }
+
         static Solution UpdateAddedDocumentToExistingContentsInSolution(
-            Solution solution, ArrayBuilder<DocumentId> addedDocumentIds)
+            Solution solution,
+            ArrayBuilder<DocumentId> addedDocumentIds,
+            Dictionary<ProjectId, ArrayBuilder<ProjectState>> projectIdToSiblingFlavors)
         {
-            ProjectId? relatedProjectIdHint = null;
             using var _ = ArrayBuilder<(DocumentId, DocumentState)>.GetInstance(out var relatedDocumentIdsAndStates);
 
-            foreach (var addedDocumentId in addedDocumentIds)
+            foreach (var group in addedDocumentIds.GroupBy(static d => d.ProjectId))
             {
-                // Ensure we don't search in addedDocumentId's project for the related document
-                if (addedDocumentId.ProjectId == relatedProjectIdHint)
-                    relatedProjectIdHint = null;
-
-                // Look for a related document we can create our contents from.  We only have to look for a single related
-                // doc as we'll be done once we update our contents to theirs.  Note: GetFirstRelatedDocumentId will also
-                // not search the project that addedDocumentId came from.  So this will help ensure we don't repeatedly add
-                // documents to a project, then look for related docs *within that project*, forcing the file-path map in it
-                // to be recreated for each document.
-                var relatedDocumentId = solution.GetFirstRelatedDocumentId(addedDocumentId, relatedProjectIdHint);
```

Comment:
Is this method used anymore?

Reply:
yes.  because we have both codepaths now.  but could potentially remove once we decide we don't need both paths.